x-pack/legacy/plugins/apm/server/lib/service_map/run_service_map_task.ts

```diff
@@ -0,0 +1,250 @@
+              }
+            },
+            {
+              'destination.address': { terms: { field: 'destination.address' } }
+            }
+          ],
+          // TODO: needs to be balanced with the 20 below
+          size: 200,
+          after: undefined
+        },
+        aggs: {
+          smpl: {
+            diversified_sampler: {
+              shard_size: 20,
+              script: {
+                lang: 'painless',
+                source: "(int)doc['span.duration.us'].value/100000"
+              }
+            },
+            aggs: {
+              tracesample: {
+                top_hits: {
+                  size: 20,
+                  _source: ['trace.id', '@timestamp']
+                }
+              }
+            }
+          }
+        }
+      }
+    }
+  };
+
+  if (afterKey) {
+    query.aggs['ext-conns'].composite.after = afterKey;
+  }
+  return query;
+}
+
+function findConns(traceIds: string[]) {
+  const query = {
+    size: 0,
+    query: {
+      bool: {
+        should: [
+          { exists: { field: 'span.id' } },
+          { exists: { field: 'transaction.type' } }
+        ] as any[],
+        minimum_should_match: 2
+      }
+    },
+    aggs: {
+      trace_id: {
+        terms: {
+          field: 'trace.id',
+          order: { _key: 'asc' },
+          size: traceIds.length
+        },
+        aggs: {
+          connections: {
+            scripted_metric: {
+              init_script: 'state.spans = new HashMap();',
+              map_script: { id: 'map-service-conns' },
+              combine_script: { id: 'combine-service-conns' },
+              reduce_script: { id: 'reduce-service-conns' }
+            }
+          }
+        }
+      }
+    }
+  };
+
+  for (const tid of traceIds) {
+    query.query.bool.should.push({ term: { 'trace.id': tid } });
+  }
+  return query;
+}
+
+interface Service {
+  name: string;
+  environment?: string;
+}
+
+interface ConnectionDoc {
+  '@timestamp': string;
+  observer: { version_major: number }; // TODO: make this dynamic
+  service: Service;
+  callee?: Service;
+  connection: Connection;
+  destination?: { address: string };
+}
+
+interface Connection {
+  upstream: { list: string[]; keyword: string };
+  in_trace: string[];
+  type: string;
+  subtype?: string;
+}
+
+export async function runServiceMapTask(
+  server: Server,
+  config: any,
+  lastRun?: string
+) {
+  const apmIdxPattern = config.get('apm_oss.indexPattern');
+  const serviceConnsDestinationIndex = config.get(
+    'xpack.apm.serviceMapDestinationIndex'
+  );
+  const serviceConnsDestinationPipeline = config.get(
+    'xpack.apm.serviceMapDestinationPipeline'
+  );
+
+  const callCluster = server.plugins.elasticsearch.getCluster('data')
+    .callWithInternalUser;
+
+  let mostRecent = '';
+  let afterKey = null;
+  while (true) {
+    const q = interestingTransactions(lastRun, afterKey);
+    const txs = await callCluster('search', {
+      index: apmIdxPattern,
```

Comment:
Shouldn't this query only search transaction indices? Currently it searches all indices (errors, spans, metrics, transactions)

Reply:
right now, it it includes connections derived from trace ids of both transactions and spans within a specific time range. we could limit this to transactions only and it may show fewer connections in the results, but eventually as more time passes and a missing trace's transaction comes within the latest time interval, those connections will be included in the service map data. I'll do what you suggest and see what happens.

Reply:
If you want to search both transactions and spans, I'd think you can specify that like:
`index: [transactionIndices, spanIndices]`